mistralrs-core/src/models/llama.rs

```diff
@@ -421,18 +423,35 @@ impl Llama {
         mut metadata: Option<(Vec<(Tensor, Tensor)>, &mut PagedAttentionInputMetadata)>,
     ) -> Result<Tensor> {
         let mut x = self.wte.forward(input_ids)?;
-        let mut cache = self.kv_cache.lock();
+        let (batch_size, seq_len, hidden_size) = x.dims3()?;
+
+        let num_devices = 1;
+        let chunk_size = seq_len / num_devices;
+
+        // let mut chunks = Vec::new();
+        // let mut chunks = Vec::<Tensor>;
+        // let chunk = x.clone();
+        // chunks.push(chunk.to_device(&self.cuda_devices[0])?);
+        let mut chunks: Vec<Tensor> = Vec::with_capacity(num_devices);
+        println!("x device {:?}", x.device());
+        chunks.push(x.copy().unwrap());
+        println!("chunk device {:?}", chunks[0].device());
+
+        let mut cache = self.kv_caches[0].lock();
         let mask = CausalMasker.make_causal_mask_as_attn_bias(
             input_ids,
             metadata
                 .as_ref()
                 .map(|(_, _)| &seqlen_offsets as &dyn PastKvLenCache)
                 .unwrap_or(&*cache as &dyn PastKvLenCache),
-            x.dtype(),
+            // x.dtype(),
+            chunks[0].dtype(),
             self.blocks[0].attn.num_attention_heads,
         )?;
         for (block_idx, block) in self.blocks.iter().enumerate() {
-            x = self.mapper.map(x, block_idx)?;
+            // x = self.mapper.map(x, block_idx)?;
+            // x = self.mapper.map(&chunks[0], block_idx)?;
+            x = self.mapper.map(chunks[0].copy().unwrap(), block_idx)?;
```

Comment:
I think this is the bit which is causing the issue. It looks like we aren't using the values from the last block as the inputs are always from the embeddings.

Reply:
So that's going to be a problem. From the algorithm in the paper, I need to use the local block on each host. So I need to iterate through the layers and then through the local blocks.

![image](https://github.com/user-attachments/assets/6dfa71b1-1fda-41e8-a19c-252226cb14df)

Any suggestions on how to redesign it?